server/cluster/metadata/cluster_metadata.go

```diff
@@ -402,6 +410,55 @@ func (c *ClusterMetadata) RegisterNode(ctx context.Context, registeredNode Regis
 	return nil
 }
 
+func needUpdate(oldCache RegisteredNode, registeredNode RegisteredNode) bool {
+	if len(oldCache.ShardInfos) != len(registeredNode.ShardInfos) {
+		return true
+	}
```

Comment:
How about moving this check to the *compare function? Otherwise, you have to add comments to explain the input shard infos must share the same length.